{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# T026 · Kinase similarity: Interaction fingerprints\n",
    "\n",
    "**Note:** This talktorial is a part of TeachOpenCADD, a platform that aims to teach domain-specific skills and to provide pipeline templates as starting points for research projects.\n",
    "\n",
    "Authors:\n",
    "\n",
    "- Dominique Sydow, 2021, [Volkamer lab, Charité](https://volkamerlab.org/)\n",
    "- Talia B. Kimber, 2021, [Volkamer lab, Charité](https://volkamerlab.org/)\n",
    "- Andrea Volkamer, 2021, [Volkamer lab, Charité](https://volkamerlab.org/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aim of this talktorial\n",
    "\n",
    "We will assess the similarity between a set of kinases based on detected protein-ligand interactions in available complex structures. The [KLIFS](https://klifs.net/) interaction fingerprint (IFP), which describes the interactions seen in a structurally resolved kinase-ligand complex, will be used in this exercise.  \n",
    "\n",
    "_Note_: We focus on similarities between orthosteric kinase binding sites; similarities to allosteric binding sites are not covered."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Contents in *Theory*\n",
    "\n",
    "* Kinase dataset\n",
    "* Kinase similarity descriptor: KLIFS interaction fingerprint\n",
    "* Fetching KLIFS data with `opencadd.databases.klifs`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Contents in *Practical*\n",
    "\n",
    "* Define the kinases of interest\n",
    "* Retrieve and preprocess data\n",
    "    * Set up a remote KLIFS session\n",
    "    * Fetch all structures describing these kinases\n",
    "    * Filter structures\n",
    "    * Fetch the structures' IFPs (if available)\n",
    "    * Merge structural and IFP data\n",
    "* Show kinase coverage\n",
    "* Compare structures\n",
    "    * Prepare IFPs as `numpy` array\n",
    "    * Calculate pairwise Jaccard distances\n",
    "* Map structure to kinase distance matrix\n",
    "* Save kinase distance matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### References\n",
    "\n",
    "* Kinase dataset: [<i>Molecules</i> (2021), <b>26(3)</b>, 629](https://www.mdpi.com/1420-3049/26/3/629) \n",
    "* Protein Data Bank\n",
    "  * PDB URL: http://www.rcsb.org/\n",
    "  * PDB database: [<i>Acta Cryst.</i> (2002), <b>D58</b>, 899-907](https://doi.org/10.1107/S0907444902003451) and [<i>Structure</i> (2012), <b>20(3)</b>, 391-396](https://doi.org/10.1016/j.str.2012.01.010)\n",
    "* KLIFS\n",
    "  * KLIFS URL: https://klifs.net/\n",
    "  * KLIFS database: [<i>Nucleic Acid Res.</i> (2020), <b>49(D1)</b>, D562-D569](https://doi.org/10.1093/nar/gkaa895)\n",
    "  * KLIFS binding site definition and interaction fingerprint calculation: [<i>J. Med. Chem.</i> (2014), <b>57(2)</b>, 249-277](https://doi.org/10.1021/jm400378w)\n",
    "* Interaction fingerprint (IFP): [<i>J. Chem. Inf. Model.</i> (2007), <b>71(1)</b>, 195-207](https://doi.org/10.1021/ci600342e)\n",
    "* `opencadd`, a Python library for structural cheminformatics\n",
    "  * GitHub repository: https://github.com/volkamerlab/opencadd\n",
    "  * Documentation: https://opencadd.readthedocs.io"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Theory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Kinase dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We use the kinase selection as defined in __Talktorial T023__."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Kinase similarity descriptor: KLIFS interaction fingerprints\n",
    "\n",
    "Interaction fingerprints (IFPs) encode the binding mode of a ligand in a binding site, i.e., the protein-ligand interactions that are present in a structurally resolved complex. If a ligand can form similar interaction patterns in proteins other than its designated protein (off- vs. on-target), it is possible that this ligand will cause unintended side effects. Knowledge about similarities between proteins can therefore help to avoid such off-target effects; or to exploit the known similarities for polypharmacology effects where one ligand could intentionally target multiple proteins at once."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this talktorial, we are interested in assessing the similarity between a set of kinases based on the similarity between the binding modes of the co-crystallized ligands. \n",
    "We will make use of the KLIFS IFP: For each kinase structure that is co-crystallized with a ligand, all interactions between the $85$ KLIFS pocket residues (see more details in __Talktorial T023__) and the ligand are described using the IFP by Marcou and Rognan ([<i>JCIM</i> (2007), <b>71(1)</b>, 195-207](https://pubs.acs.org/doi/10.1021/ci600342e))."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is how the IFP is explained in the respective KLIFS publication [<i>J. Med. Chem.</i> (2014), <b>57(2)</b>, 249-277](https://doi.org/10.1021/jm400378w):\n",
    "\n",
    "> For each amino acid in the catalytic cleft, seven types of protein−ligand interactions are determined. The presence of a certain type of interaction results in the type-setting of a “1” in the bit-string; otherwise a “0” is used to indicate the absence of the interaction. The following seven types of interactions are summarized in one bit-string; bit 1 = hydrophobic contact (HYD); bit 2 = face to face aromatic interactions (F−F); bit 3 = face to edge aromatic interactions (F−E); bit 4 = protein H-bond donor (DON); bit 5 = protein H- bond acceptor (ACC); bit 6 = protein cationic interactions (ION+), and bit 7 = protein anionic interactions (ION−)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This results in a $85 \\times 7 = 595$ bit vector. Since the binding site is aligned across all kinases, each bit position in this IFP can be directly compared across all IFPs in KLIFS. This is what we will do in the practical part of this tutorial."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![KLIFS IFP](images/T026_KLIFS_IFP.png)\n",
    "\n",
    "*Figure 1:* \n",
    "KLIFS interaction fingerprint (IFP) definition: Seven interaction types are detected between each of a kinase structure's $85$ pocket residues and a co-crystallized ligand. Interaction types include: hydrophobic contact (HYD), face to face aromatic interactions (F−F), face to edge aromatic interactions (F−E), protein H-bond donor (DON), protein H-bond acceptor (ACC), protein cationic interactions (ION+), and protein anionic interactions (ION−).\n",
    "Figure by Dominique Sydow, adapted from: [<i>J. Med. Chem.</i> (2014), <b>57(2)</b>, 249-277](https://doi.org/10.1021/jm400378w)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fetching KLIFS data with `opencadd.databases.klifs`\n",
    "\n",
    "`opencadd` is a Python library for structural cheminformatics developed by the [Volkamer lab](https://volkamerlab.org/) at the Charité in Berlin. \n",
    "This library is a growing collection of modules that help facilitate and standardize common tasks in structural bioinformatics and cheminformatics. Today, we will use the module `opencadd.databases.klifs`, which allows us to fetch the KLIFS IFP as a `pandas` DataFrame.\n",
    "\n",
    "- GitHub repository: https://github.com/volkamerlab/opencadd\n",
    "- Documentation: https://opencadd.readthedocs.io\n",
    "\n",
    "For more information about this library and the KLIFS OpenAPI, please refer to __Talktorial T012__."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Practical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from sklearn.metrics import pairwise_distances\n",
    "from scipy.spatial import distance\n",
    "from opencadd.databases.klifs import setup_remote"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "HERE = Path(_dh[-1])\n",
    "DATA = HERE / \"data\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Run in demo mode: True\n"
     ]
    }
   ],
   "source": [
    "configs = pd.read_csv(HERE / \"../T023_what_is_a_kinase/data/pipeline_configs.csv\")\n",
    "configs = configs.set_index(\"variable\")[\"default_value\"]\n",
    "\n",
    "DEMO = bool(int(configs[\"DEMO\"]))\n",
    "N_STRUCTURES_PER_KINASE = int(configs[\"N_STRUCTURES_PER_KINASE\"])\n",
    "N_CORES = int(configs[\"N_CORES\"])\n",
    "\n",
    "print(f\"Run in demo mode: {DEMO}\")\n",
    "if not DEMO:\n",
    "    if N_STRUCTURES_PER_KINASE > 0:\n",
    "        print(f\"Number of structures per kinase: {N_STRUCTURES_PER_KINASE}\")\n",
    "    else:\n",
    "        print(f\"Number of structures per kinase: all available structures\")\n",
    "    print(f\"Number of cores used: {N_CORES}\")\n",
    "\n",
    "# NBVAL_CHECK_OUTPUT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define the kinases of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's load the kinase selection as defined in __Talktorial T023__."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>kinase</th>\n",
       "      <th>kinase_klifs</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>group</th>\n",
       "      <th>full_kinase_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>P00533</td>\n",
       "      <td>TK</td>\n",
       "      <td>Epidermal growth factor receptor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ErbB2</td>\n",
       "      <td>ErbB2</td>\n",
       "      <td>P04626</td>\n",
       "      <td>TK</td>\n",
       "      <td>Erythroblastic leukemia viral oncogene homolog 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PI3K</td>\n",
       "      <td>p110a</td>\n",
       "      <td>P42336</td>\n",
       "      <td>Atypical</td>\n",
       "      <td>Phosphatidylinositol-3-kinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VEGFR2</td>\n",
       "      <td>KDR</td>\n",
       "      <td>P35968</td>\n",
       "      <td>TK</td>\n",
       "      <td>Vascular endothelial growth factor receptor 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>P15056</td>\n",
       "      <td>TKL</td>\n",
       "      <td>Rapidly accelerated fibrosarcoma isoform B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CDK2</td>\n",
       "      <td>CDK2</td>\n",
       "      <td>P24941</td>\n",
       "      <td>CMGC</td>\n",
       "      <td>Cyclic-dependent kinase 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>LCK</td>\n",
       "      <td>LCK</td>\n",
       "      <td>P06239</td>\n",
       "      <td>TK</td>\n",
       "      <td>Lymphocyte-specific protein tyrosine kinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MET</td>\n",
       "      <td>MET</td>\n",
       "      <td>P08581</td>\n",
       "      <td>TK</td>\n",
       "      <td>Mesenchymal-epithelial transition factor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>p38a</td>\n",
       "      <td>p38a</td>\n",
       "      <td>Q16539</td>\n",
       "      <td>CMGC</td>\n",
       "      <td>p38 mitogen activated protein kinase alpha</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   kinase kinase_klifs uniprot_id     group  \\\n",
       "0    EGFR         EGFR     P00533        TK   \n",
       "1   ErbB2        ErbB2     P04626        TK   \n",
       "2    PI3K        p110a     P42336  Atypical   \n",
       "3  VEGFR2          KDR     P35968        TK   \n",
       "4    BRAF         BRAF     P15056       TKL   \n",
       "5    CDK2         CDK2     P24941      CMGC   \n",
       "6     LCK          LCK     P06239        TK   \n",
       "7     MET          MET     P08581        TK   \n",
       "8    p38a         p38a     Q16539      CMGC   \n",
       "\n",
       "                                   full_kinase_name  \n",
       "0                  Epidermal growth factor receptor  \n",
       "1  Erythroblastic leukemia viral oncogene homolog 2  \n",
       "2                     Phosphatidylinositol-3-kinase  \n",
       "3     Vascular endothelial growth factor receptor 2  \n",
       "4        Rapidly accelerated fibrosarcoma isoform B  \n",
       "5                         Cyclic-dependent kinase 2  \n",
       "6       Lymphocyte-specific protein tyrosine kinase  \n",
       "7          Mesenchymal-epithelial transition factor  \n",
       "8        p38 mitogen activated protein kinase alpha  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kinase_selection_df = pd.read_csv(HERE / \"../T023_what_is_a_kinase/data/kinase_selection.csv\")\n",
    "kinase_selection_df\n",
    "# NBVAL_CHECK_OUTPUT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Retrieve and preprocess data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we query the KLIFS database using the `opencadd.databases.klifs` module to generate our IFP dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Set up a remote KLIFS session"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from opencadd.databases.klifs import setup_remote"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "klifs_session = setup_remote()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fetch all structures describing these kinases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of structures: 2523\n",
      "Kinases: CDK2 p38a EGFR ErbB2 MET LCK KDR BRAF p110a\n"
     ]
    }
   ],
   "source": [
    "# Get list of kinase names\n",
    "kinase_names = kinase_selection_df[\"kinase_klifs\"].to_list()\n",
    "\n",
    "# Get all available structures for these kinases\n",
    "structures_df = klifs_session.structures.by_kinase_name(kinase_names=kinase_names)\n",
    "# Keep only relevant columns\n",
    "structures_df = structures_df[\n",
    "    [\n",
    "        \"structure.klifs_id\",\n",
    "        \"kinase.klifs_name\",\n",
    "        \"species.klifs\",\n",
    "        \"structure.dfg\",\n",
    "        \"structure.resolution\",\n",
    "        \"structure.qualityscore\",\n",
    "    ]\n",
    "]\n",
    "print(f\"Number of structures: {len(structures_df)}\")\n",
    "print(\"Kinases:\", *structures_df[\"kinase.klifs_name\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Filter structures\n",
    "\n",
    "We filter the structures by different criteria:\n",
    "\n",
    "- Species: human\n",
    "- Conformation: DFG-in (the active kinase conformation)\n",
    "- Resolution: $\\le 3$ Angström\n",
    "- Quality score*: $\\ge 6$\n",
    "\n",
    "\\* The KLIFS quality score takes into account the quality of the alignment, as well as the number of missing residues and atoms. A higher score indicates a better structure quality."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of structures: 1657\n",
      "Kinases: CDK2 p38a EGFR ErbB2 MET LCK KDR BRAF p110a\n"
     ]
    }
   ],
   "source": [
    "structures_df = structures_df[\n",
    "    (structures_df[\"species.klifs\"] == \"Human\")\n",
    "    & (structures_df[\"structure.dfg\"] == \"in\")\n",
    "    & (structures_df[\"structure.resolution\"] <= 3)\n",
    "    & (structures_df[\"structure.qualityscore\"] >= 6)\n",
    "]\n",
    "print(f\"Number of structures: {len(structures_df)}\")\n",
    "print(\"Kinases:\", *structures_df[\"kinase.klifs_name\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the structure KLIFS IDs for the next step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of structures: 1657\n"
     ]
    }
   ],
   "source": [
    "structure_klifs_ids = structures_df[\"structure.klifs_id\"].to_list()\n",
    "print(f\"Number of structures: {len(structure_klifs_ids)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_Note for demo mode_: To make it easier for us to maintain the talktorials, we will now load a set of frozen structure KLIFS IDs (2021-08-23) and continue to work with those.\n",
    "\n",
    "_Note for non-demo mode_: Did you specify `N_STRUCTURES_PER_KINASE` in the configuration file? If you e.g. set a value of 3, we will select in the following the top 3 structures per kinase in terms of resolution and KLIFS quality score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notebook is run in demo mode - load frozen structure set.\n",
      "Number of structures: 1620\n"
     ]
    }
   ],
   "source": [
    "if DEMO:\n",
    "    # Load frozen dataset\n",
    "    print(\"Notebook is run in demo mode - load frozen structure set.\")\n",
    "    structure_klifs_ids = pd.read_csv(DATA / \"frozen_structure_klifs_ids.csv\")[\n",
    "        \"structure.klifs_id\"\n",
    "    ].to_list()\n",
    "    structures_df = structures_df[\n",
    "        structures_df[\"structure.klifs_id\"].isin(structure_klifs_ids)\n",
    "    ].copy()\n",
    "else:\n",
    "    if N_STRUCTURES_PER_KINASE > 0:\n",
    "        print(f\"Select {N_STRUCTURES_PER_KINASE} structures per kinase for downstream analysis.\")\n",
    "        # Sort structures by kinase and quality\n",
    "        structures_df = structures_df.sort_values(\n",
    "            by=[\"kinase.klifs_name\", \"structure.resolution\", \"structure.qualityscore\"],\n",
    "            ascending=[True, True, False],\n",
    "        )\n",
    "        # Reduce number of structures per kinase\n",
    "        structures_df = structures_df.groupby(\"kinase.klifs_name\").head(N_STRUCTURES_PER_KINASE)\n",
    "        structure_klifs_ids = structures_df[\"structure.klifs_id\"].to_list()\n",
    "    else:\n",
    "        print(f\"Use all available structures per kinase for downstream analysis.\")\n",
    "\n",
    "print(f\"Number of structures: {structures_df.shape[0]}\")\n",
    "# NBVAL_CHECK_OUTPUT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Fetch the structures' IFPs (if available)\n",
    "\n",
    "We fetch the IFPs for the set of structures. Not all structures will have an IFP because not all structures have a co-crystallized ligand."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of structures with IFPs: 1466\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>structure.klifs_id</th>\n",
       "      <th>interaction.fingerprint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>775</td>\n",
       "      <td>0000000000000010000000000000000000000000000000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>777</td>\n",
       "      <td>0000000000000010000001000000000000000000000000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>778</td>\n",
       "      <td>0000000000000010000000000000000000000000000000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>779</td>\n",
       "      <td>0000000000000010000001000000000000000000000000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>782</td>\n",
       "      <td>0000000000000010000001000000000000000000000000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   structure.klifs_id                            interaction.fingerprint\n",
       "0                 775  0000000000000010000000000000000000000000000000...\n",
       "1                 777  0000000000000010000001000000000000000000000000...\n",
       "2                 778  0000000000000010000000000000000000000000000000...\n",
       "3                 779  0000000000000010000001000000000000000000000000...\n",
       "4                 782  0000000000000010000001000000000000000000000000..."
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ifps_df = klifs_session.interactions.by_structure_klifs_id(structure_klifs_ids)\n",
    "print(f\"Number of structures with IFPs: {len(ifps_df)}\")\n",
    "ifps_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge structural and IFP data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of structures with IFPs: 1466\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>structure.klifs_id</th>\n",
       "      <th>interaction.fingerprint</th>\n",
       "      <th>kinase.klifs_name</th>\n",
       "      <th>species.klifs</th>\n",
       "      <th>structure.dfg</th>\n",
       "      <th>structure.resolution</th>\n",
       "      <th>structure.qualityscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>775</td>\n",
       "      <td>0000000000000010000000000000000000000000000000...</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Human</td>\n",
       "      <td>in</td>\n",
       "      <td>3.00</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>777</td>\n",
       "      <td>0000000000000010000001000000000000000000000000...</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Human</td>\n",
       "      <td>in</td>\n",
       "      <td>2.64</td>\n",
       "      <td>6.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>778</td>\n",
       "      <td>0000000000000010000000000000000000000000000000...</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Human</td>\n",
       "      <td>in</td>\n",
       "      <td>1.90</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>779</td>\n",
       "      <td>0000000000000010000001000000000000000000000000...</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Human</td>\n",
       "      <td>in</td>\n",
       "      <td>2.80</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>782</td>\n",
       "      <td>0000000000000010000001000000000000000000000000...</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>Human</td>\n",
       "      <td>in</td>\n",
       "      <td>1.70</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   structure.klifs_id                            interaction.fingerprint  \\\n",
       "0                 775  0000000000000010000000000000000000000000000000...   \n",
       "1                 777  0000000000000010000001000000000000000000000000...   \n",
       "2                 778  0000000000000010000000000000000000000000000000...   \n",
       "3                 779  0000000000000010000001000000000000000000000000...   \n",
       "4                 782  0000000000000010000001000000000000000000000000...   \n",
       "\n",
       "  kinase.klifs_name species.klifs structure.dfg  structure.resolution  \\\n",
       "0              EGFR         Human            in                  3.00   \n",
       "1              EGFR         Human            in                  2.64   \n",
       "2              EGFR         Human            in                  1.90   \n",
       "3              EGFR         Human            in                  2.80   \n",
       "4              EGFR         Human            in                  1.70   \n",
       "\n",
       "   structure.qualityscore  \n",
       "0                     8.0  \n",
       "1                     6.8  \n",
       "2                     8.0  \n",
       "3                     8.0  \n",
       "4                     8.0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "structures_with_ifps_df = ifps_df.merge(structures_df, on=\"structure.klifs_id\", how=\"inner\")\n",
    "print(f\"Number of structures with IFPs: {len(structures_with_ifps_df)}\")\n",
    "structures_with_ifps_df.head()\n",
    "# NBVAL_CHECK_OUTPUT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Show kinase coverage\n",
    "\n",
    "Let's get the number of structures that describe our kinases (kinase coverage)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "kinase.klifs_name\n",
       "ErbB2      4\n",
       "KDR        6\n",
       "LCK       30\n",
       "p110a     45\n",
       "BRAF      57\n",
       "MET       95\n",
       "p38a     127\n",
       "EGFR     339\n",
       "CDK2     763\n",
       "dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Use pandas' groupby method to count the number of structures (rows) per kinase\n",
    "n_structures_per_kinase = structures_with_ifps_df.groupby(\"kinase.klifs_name\").size().sort_values()\n",
    "n_structures_per_kinase\n",
    "# NBVAL_CHECK_OUTPUT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's plot the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZoAAAEGCAYAAABcolNbAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAAoq0lEQVR4nO3de5xVdb3/8debGUERhZIgBHXCSJDbqARqHiUJQ+NUpImIHhTUsuMpTS2sc47Wr46kXTTpVJgXjnLELG9HEFPBMFNxlM3FlEhFQVG8onERGD6/P9Ya2ozDsIFZe++ZeT8fj/2Yvb7ru9Z670H8sL5r7e9SRGBmZpaVNqUOYGZmLZsLjZmZZcqFxszMMuVCY2ZmmXKhMTOzTFWWOkCxde7cOaqqqkodw8ys2ejcuTP33XfffRExYme2b3WFpqqqipqamlLHMDNrViR13tltPXRmZmaZcqExM7NMudCYmVmmXGjMzCxTre5mgEUvr6Zq4oxSxzAzK6plkz5XsmP7jMbMzDLlQmNm1gosWbKE6urqLa+9996bq666CoBrrrmGgw46iL59+/Ktb30LgHnz5m3pO3DgQIBOO3vsog6dSfoocBXwSeB9YBlwPrAAeBbYHXgP+EVETE23OQMYFBHnSWoD3ADUAv8G/BY4MF3+v4iYWLxPY2bWfBx00EHkcjkAamtr6d69O6NGjWLOnDncddddLFy4kHbt2rFq1SoA+vXrR01NDZWVlaxcuZJ99933AEmVEbFpR49dtEIjScAdwNSIOCVtqwa6As9FxCFpW0/gdkltIuKGetv/CtgNOJOkKP04IuZIags8KOn4iLi3WJ/JzKw5evDBBznwwAM54IADuPjii5k4cSLt2rUDoEuXLgC0b99+S//169fv0vGKOXT2aWBjRPyqriEicsDy/E4R8TzwTeDr9ba/GtgH+JeI2BwRayNiTrrNBuApoEd28c3MWobp06czZswYAP7617/y8MMPM2TIEI455hieeOKJLf0ef/xx+vbtS//+/QFe3JmzGShuoekHPFlg36eA3nnLpwKHAac09EEldQL+GXiwoZ1JOkdSjaSa2rWrdyi0mVlLsmHDBu6++26+/OUvA7Bp0ybefvttHnvsMa688kpOPvlk6p68PGTIEJ5++um64tNN0u47c8xyvRlA9ZafAg4ABn+go1QJ3AL8PD0b+oCImBIRgyJiUEX7jk0e1sysubj33ns59NBD6dq1KwA9evTgS1/6EpIYPHgwbdq04Y033thqmz59+kByLbzfzhyzmIXmaZKzkkIcAjyTt/wscDJwq6S+9fpOAZZGxFW7nNDMrIW75ZZbtgybAXzxi19k9uzZQDKMtmHDBjp37swLL7zApk3JANKLL74IyXXxZTtzzGIWmtlAO0ln1zVI+iTJmQp5bVXAj4Fr8tsj4s/AV4EZkvZP+/4A6Ehy55qZmTVi7dq13H///XzpS1/a0jZ+/Hief/55+vXrxymnnMLUqVORxJ/+9CcGDhxIdXU1o0aNAngpIt7Y5s4bobqxuGKQtC/J7c2HAev5x+3NC9n69uZf1t1xln97c7p8JnAxcHy6/bMkt0oDTI6I3zSWoV23XtFt3FVN9pnMzJqDXZ0ZQNKTETFoZ7Yt6vdoIuIVkiGw+vZoZJsbgRvzlm8g+S4NfPBajpmZlZlWN9dZ/+4dqSnhnD9mZq1Nud51ZmZmLYQLjZmZZcqFxszMMuVCY2ZmmXKhMTOzTLnQmJlZplxozMwsUy40ZmaWKRcaMzPLlAuNmZllyoXGzMwy5UJjZmaZanWTai56eTVVE2eUOoY1I7s6vbpZa+czGrMCrF+/nsGDBzNw4ED69u3LpZdeCsB//Md/MGDAAKqrqznuuON45ZVXgOS57GeeeSb9+/dn4MCBPPTQQyVMb1ZaRS00kmol5fJeE9P2Skn/JWlp3rrvNrJdlaShklZLmi/pWUk/LuZnsdalXbt2zJ49mwULFpDL5Zg1axaPPfYYF198MQsXLiSXyzFy5Ei+//3vA3DttdcCsGjRIu6//34uvPBCNm/eXMqPYFYyxR46WxcR1Q20/wD4KNA/ItZL2gu4sLHt0kc+PxwRIyXtAcyXdEdEPJJNdGvNJNGhQwcANm7cyMaNG5HE3nvvvaXPmjVrkJJn8f3lL39h2LBhAHTp0oVOnTpRU1PD4MGDix/erMRKPnQmqT1wNvBvEbEeICLei4jLCt1HRKwDckD3LDKaAdTW1lJdXU2XLl0YPnw4Q4YMAeC73/0u++23H9OmTdtyRjNw4EDuuusuNm3axAsvvMCTTz7J8uXLSxnfrGSKXWj2qDcENhr4OPBSRLxX4HZ31F8p6UNAL2BuRrnNqKioIJfLsWLFCubNm8fixYsB+OEPf8jy5csZO3YskydPBmD8+PH06NGDQYMGcf7553PkkUdSWdnq7r0xA4pfaNZFRHXe69b6HSSdmRaU5ZL2a2C7UXnd/0nSQuBV4J6IeLWhg0o6R1KNpJrataub/lNZq9KpUyeGDh3KrFmztmo/9dRT+f3vfw9AZWUlP/vZz8jlctx1112888479OrVqxRxzUqu5ENnwN+A/dPrMkTEDen1mNVAxXa2fTgiBgD9gXMlVTfUKSKmRMSgiBhU0b5j0yW3VuP111/nnXfeAWDdunU88MAD9O7dm6VLl27pc/fdd9O7d28A1q5dy5o1awC4//77qays5OCDDy56brNyUPJz+YhYK+k6YLKkr6Q3A1QAbXdgH3+VdDnwbWBMVlmt9Vq5ciXjxo2jtraWzZs3c/LJJzNy5EhOPPFElixZQps2bTjggAP41a9+BcCqVav47Gc/S5s2bejevTs33XRTiT+BWekUu9DsISmXtzwrIiYC3wX+H7BY0nvAOmAq8MoO7PtXwEWSPhYRLzRVYDOAAQMGMH/+/A+01w2V1VdVVcWSJUuyjmXWLBS10EREg0NhEbERmJi+GlrfoYG2h4CH8pbX4bvOzMzKTjlcozEzsxas5Ndoiq1/947UeO4qM7Oi8RmNmZllyoXGzMwy5UJjZmaZcqExM7NMudCYmVmmXGjMzCxTLjRmZpYpFxozM8uUC42ZmWXKhcbMzDLlQmNmZplqdXOdLXp5NVUTZ5Q6RpNa5rnbzKyM+YzGzMwy5ULTQowfP54uXbrQr1+/LW0XX3wxvXv3ZsCAAYwaNWrLo4inTZtGdXX1llebNm3I5XKlCW5mLV7ZFBpJ10laIGmhpN9J6pC2d5T0f+m6pyWdWeqs5eiMM85g1qxZW7UNHz6cxYsXs3DhQj7xiU9w+eWXAzB27FhyuRy5XI6bbrqJqqoqqqurS5DazFqDsik0wAURMTAiBgAvAeel7f8K/CUiBgJDgZ9IaluijGXr6KOP5sMf/vBWbccddxyVlclluMMPP5wVK1Z8YLtbbrmFMWPGFCWjmbVORS80kqokPStpat7ZS/uIeDddL2APINJNAtgrbe8AvAVsSvveKenJ9EznnGJ/lubk+uuv5/jjj/9A+6233upCY2aZKtUZzUHAlPTs5V3gawCSbgBeBXoD16R9JwN9gFeARcA3ImJzum58RBwGDAK+Lmmfhg4m6RxJNZJqateuzuozla0f/vCHVFZWMnbs2K3aH3/8cdq3b7/VdR0zs6ZWqkKzPCIeSd/fDBwFEBFnAvsCzwCj0/WfBXJpezUwWdLe6bqvS1oAPAbsB/Rq6GARMSUiBkXEoIr2HZv+05SxqVOncs899zBt2jSSk8J/mD59us9mzCxzpfoeTWxrOSJqJd0KXAzcAJwJTIqIAP4m6QWgt6T2wGeAIyJiraSHgN2Lkr6ZmDVrFj/60Y/44x//SPv27bdat3nzZm677Tbmzp1bonRm1lqU6oxmf0lHpO/HAH+S9HHYco3mn4Fn0/UvAcPSdV1Jht2eBzoCb6dFpjdweBHzl50xY8ZwxBFHsGTJEnr06MF1113Heeedx3vvvcfw4cOprq7mq1/96pb+c+fOpUePHvTs2bOEqc2sNVByolDEA0pVwExgLnAksBQ4Hbgf2BsQsAA4NyLelbQvcCPQLV03KSJultQOuBPoDiwBPgJcFhEPNXb8dt16RbdxVzX1xyopzwxgZlmT9GREDNqZbUs1dLY5Ir5ar+1TDXWMiFeA4xpofx/44G1UZmZWVlrdXGf9u3ekxmcAZmZFU/RCExHLAN9Pa2bWSpTTzABmZtYCudCYmVmmXGjMzCxTLjRmZpYpFxozM8uUC42ZmWXKhcbMzDLlQmNmZplyoTEzs0wVXGgk7ZllEDMza5m2OwWNpCOB35A8Rnl/SQOBr0TE17IOl4VFL6+mauKMUsdokGdhNrOWqJAzmp+RPOXyTYCIWAAcnWUoMzNrOQoaOouI5fWaajPIYqmrr76afv360bdvX6666ioALrvsMrp37051dTXV1dXMnDmztCHNzApUyOzNy9Phs5DUFvg68ExThpAUwM0RcXq6XAmsBB6PiJGSzgCuBF7O22wcMDV9vz+wOn29ERGfacp8xbR48WKuvfZa5s2bR9u2bRkxYgSf+1wypHbBBRdw0UUXlTihmdmOKaTQfBW4muRJliuAPwD/2sQ51gD9JO0REeuA4WxdVABujYjz6rVVA0i6EbgnIn7XxLmK7plnnuHwww+nffv2ABxzzDHccccdJU5lZrbztjt0FhFvRMTYiOgaEV0i4rSIeDODLPcCdVfDxwC3ZHCMstevXz/mzp3Lm2++ydq1a5k5cybLlycjl5MnT2bAgAGMHz+et99+u8RJzcwKs91CI+ljkn4q6XZJd9e9MsgyHThF0u7AAODxeutHS8rlvfYodMeSzpFUI6mmdu3qpszc5Pr06cO3v/1thg8fzogRIxg4cCCVlZWce+65PPfcc+RyObp168aFF15Y6qhmZgUpZOjsTuA64P+AzVkFiYiFkqpIzmYautLd0NBZofueAkwBaNetV+x0yCKZMGECEyZMAOA73/kOPXr0oGvXrlvWn3322YwcObJU8czMdkghhWZ9RPw88ySJu4EfA0OBfYp0zLKzatUqunTpwksvvcTtt9/Oo48+ysqVK+nWrRsAd9xxB/36+WnYZtY8FFJorpZ0KclNAO/XNUbEUxnkuR5YHRGLJA3NYP/Nwoknnsibb77Jbrvtxi9+8Qs+9KEPcfrpp5PL5ZBEVVUVv/71r0sd08ysIIUUmv7A6cCx/GPoLNLlJhURK0jucGvIaElH5S1/LSL+3NQZysHDDz/8gbabbrqpBEnMzHZdIYVmFNAzIjZkFSIiOjTQ9hDwUPr+RuDGRrY/I5NgZma2ywopNAuATsCqbKMUR//uHanxnGJmZkVTSKHpCjwr6Qm2vkbz+cxSmZlZi1FIobk08xRmZtZibbfQRMQfixHEzMxapkJmBjhc0hOS/i5pg6RaSe8WI5yZmTV/hTwmYDLJt/WXAnsAZ6VtZmZm21XINRoi4m+SKiKiFrhBUov8/oqZmTW9QgrN2vQ5NDlJV5A8J2bPbGOZmVlLUcjQ2elpv/NInhuzH3BilqHMzKzlKOSusxfTt+uB72Ubx8zMWprtFhpJnwIuAw7I7x8RPbOLZWZmLUUh12iuAy4AngRqs41jZmYtTSGFZnVE3Jt5kiJZ9PJqqibOyGz/yzyPmpnZVgopNHMkXQncTvbPozEzsxamkEIzJP05KK8tk+fRtBRVVVXstddeVFRUUFlZSU1NDaNHj2bJkiUAvPPOO3Tq1IlcLlfaoGZmRVDIXWefbsoDSqoFFgEiueZzXkT8WVIV8AywBGgL1AATImJjul0l8CpwbURckre/h4BuwLq06QcR8bumzLwz5syZQ+fOnbcs33rrrVveX3jhhXTs2LEUsczMiq6gmQGa2LqIqAaQ9FngcuCYdN1zEVEtqQK4HzgZmJauO46kCJ0s6TsREXn7HBsRNUVJv4sigt/+9rfMnj271FHMzIqikC9sZmlv4O36jelUN/OA7nnNY0ge8/wScHhR0u0kSRx33HEcdthhTJkyZat1Dz/8MF27dqVXr14lSmdmVlylOKPZQ1IO2J1kyOsD13ok7U5ybegb6fIewDDgKyRP+xwDPJq3yTRJdUNnwyLizXr7Owc4B6Bi74804Udp2COPPMK+++7LqlWrGD58OL179+boo48G4JZbbmHMmDGZZzAzKxeFPCagvaT/kHRtutxL0shdOOa6iKiOiN7ACOB/JCldd2BahN4EXoqIhWn7SGBORKwFfg+MSofX6oxN91ldv8gARMSUiBgUEYMq2md/bWTfffcFoEuXLowaNYp58+YBsGnTJm6//XZGjx6deQYzs3JRyNDZDSS3NR+RLq8AftAUB4+IR4HOQN1pxnPp9ZuPA4dLqntc9BjgM5KWkXxxdB+gSW9SaCpr1qzhvffe2/L+D3/4A/369QPggQceoHfv3vTo0aOUEc3MiqqQobMDI2K0pDEAEbEu7wxkl0jqDVSQnMG0r2uPiJWSJgKXpHeVHQXsFxHvp9udSVJ8HmiKHE3ptddeY9SoUUByBnPqqacyYsQIAKZPn+5hMzNrdQopNBvSayQBIOlA8r64uRPqrtFAcovzuIiobaB23Ukyx9o3gNl1RSZ1F3CFpHa7kCMTPXv2ZMGCBQ2uu/HGG4sbxsysDBRSaC4FZgH7SZoGfAo4Y2cPGBEV22hfBvTLWw5g4Db6vsU/htuG7mwWMzPLnrb+Oso2Okn7kNxSLOCxiHgj62BZGTRoUNTUNIuv3JiZlQ1JT0bEoO33/KBC7jr7FLA+ImaQ3Fr8HUkH7MzBzMys9SnkrrNfkjzOeSBwMfAi8D+ZpjIzsxajkEKzKb1e8gXg5xFxNbBXtrHMzKylKORmgPckXQKcBhydflFyt2xjmZlZS1HIGc1oktuZJ0TEqyTzj12ZaSozM2sxCnlMwKvAT/OWX8LXaMzMrECF3HV2uKQnJP1d0gZJtZJWFyOcmZk1f4UMnU0mme5lKbAHcBbwiyxDmZlZy1HQYwIi4m+SKtLnxNwg6c8Z5zIzsxaikEKzVlJbICfpCmAlsGe2sczMrKUoZOjs9LTfecAaYD/gxCxDmZlZy1HQXGctSbtuvaLbuKuaZF/LJn2uSfZjZlbudmWus+0OnaVznV0GHJDfPyJ67swBzcysdSlk6Ow6ku/RHAV8Mu9lqdraWg455BBGjkyecH3ZZZfRvXt3qqurqa6uZubMmSVOaGZWOoUUmtURcW9ErIqIN+teO3tASedJ+pukkNQ5r723pEclvS/ponrbjJC0JN1u4s4eOytXX301ffr02artggsuIJfLkcvlOOGEE0qUzMys9AopNHMkXSnpCEmH1r124ZiPAJ8hmQU631vA14Ef5zemc6v9AjgeOBgYI+ngXTh+k1qxYgUzZszgrLPOKnUUM7OyVEihGQIMAv4L+En6+nGjWwCSqiQ9K2mqpIWSfiepfUTMT5+muZX0jOkJYGO9VYOBv0XE8xGxAZhOMpM0ks5OZy1YIOn3ktoX8Hma1Pnnn88VV1xBmzZb/yonT57MgAEDGD9+PG+//XaxY5mZlY3tFpqI+HQDr2ML3P9BwJSIGAC8C3xtJzJ2B5bnLa9I2wBuj4hPRsRA4BlgQkM7kHSOpBpJNbVrm272nHvuuYcuXbpw2GGHbdV+7rnn8txzz5HL5ejWrRsXXnhhkx3TzKy52eZdZ5JOi4ibJX2zofUR8dOG2utZHhGPpO9vpoGhsQKoocOnP/tJ+gHJkz87APc1tIOImAJMgeT25h08/jY98sgj3H333cycOZP169fz7rvvctppp3HzzTdv6XP22WdvuUnAzKw1auyMpu7b/3s18OpQ4P7r/099Z/4nv4LkS6J1egCvpO9vBM6LiP7A94Ddd2L/O+3yyy9nxYoVLFu2jOnTp3Psscdy8803s3Llyi197rjjDvr161fMWGZmZaWx79HMAIiI79VfIemfC9z//pKOiIhHSSbm/NOOR+QJoJekjwEvA6cAp6br9gJWStoNGJuuL7lvfetb5HI5JFFVVcWvf/3rUkcyMyuZbc4MIGkJ8Nn6F+4lnQn8e0Qc2OiOpSpgJjAXOJJk9ufTSWZ//hbwUWAVMDMizpL0UaAG2BvYDPwdODgi3pV0AnAVUAFcHxE/TI9xbrqvF4FFwF4RcUZjuTwzgJnZjstqZoALgPslnRARS9MDXUJyNnFMgfvfHBFfrdf28/S1lfQBaz0a2klEzCQpWvXbfwn8ssAsZmZWAtssNBExU9L7wL2SvkhyJvJJ4OiIaLb36/bv3pEan4mYmRVNo7c3R8SDwBnAQ0BPYFihRSYilkWEr4KbmbVyjd3e/B7JXWIC2gHDgFWSBERE7F2ciGZm1pw1NnS2VzGDmJlZy1TIFDRmZmY7zYXGzMwy5UJjZmaZcqExM7NMudCYmVmmXGjMzCxTLjRmZpYpFxozM8tUY5NqtkiLXl5N1cQZBff3DM1mZrvGZzRmZpYpF5oCrF+/nsGDBzNw4ED69u3LpZdeCsBbb73F8OHD6dWrF8OHD+ftt5vtpNZmZpkpq0Ij6e/baP8XSYslPS3pL5IuSttvlHRS+v7DkuanD2ZrUu3atWP27NksWLCAXC7HrFmzeOyxx5g0aRLDhg1j6dKlDBs2jEmTJjX1oc3Mmr2yKjQNkXQ8cD5wXET0BQ4FVtfr0xG4D5gSETdkkIEOHToAsHHjRjZu3Igk7rrrLsaNGwfAuHHjuPPOO5v60GZmzV7ZFxrgEuCiiHgFICLWR8S1ees7APcC/5s+cTMTtbW1VFdX06VLF4YPH86QIUN47bXX6NatGwDdunVj1apVWR3ezKzZag6Fph/wZCPrfwr8KSJ+tq0Oks6RVCOppnbt6m11a1RFRQW5XI4VK1Ywb948Fi9evFP7MTNrbZpDodme2cAXJHXZVoeImBIRgyJiUEX7jrt0sE6dOjF06FBmzZpF165dWblyJQArV66kS5dtRjAza7WaQ6F5GjiskfXTgV8CMyVl8rC2119/nXfeeQeAdevW8cADD9C7d28+//nPM3XqVACmTp3KF77whSwOb2bWrDWHL2xeDlwhaWREvCqpHfCViPh5XYeIuEpSN+AOSSdExIamDLBy5UrGjRtHbW0tmzdv5uSTT2bkyJEcccQRnHzyyVx33XXsv//+3HbbbU15WDOzFqHcCk17SSvyln8aET+V1BV4QJKAAK6vv2FEfFvSDcBNksZExOamCjVgwADmz5//gfZ99tmHBx98sKkOY2bWIpVVoYmIBofy0luWP3DbckScUW+5yb9DY2Zmu6asCk0x9O/ekRrPX2ZmVjTN4WYAMzNrxlxozMwsUy40ZmaWKRcaMzPLlAuNmZllyoXGzMwy5UJjZmaZcqExM7NMudCYmVmmXGjMzCxTLjRmZpYpFxozM8tUqys0i15eTdXEGaWOYWbWarS6QlPfO++8w0knnUTv3r3p06cPjz76aKkjmZm1KGVRaCT9Pe/9CZKWStpf0mWSXpaUS9tul3RwXt+HJC2RtEDSE5Kqd/TY3/jGNxgxYgTPPvssCxYsoE+fPk30qczMDMqk0NSRNAy4BhgRES+lzT+LiOqI6AXcCsyW9JG8zcZGxEDgv4Erd+R47777LnPnzmXChAkAtG3blk6dOu3qxzAzszxlU2gk/RNwLfC5iHiuoT4RcSvwB+DUBlY/CnTfkWM+//zzfOQjH+HMM8/kkEMO4ayzzmLNmjU7Gt3MzBpRLoWmHXAX8MWIeHY7fZ8CejfQPgK4c0cOumnTJp566inOPfdc5s+fz5577smkSZN2ZBdmZrYd5VJoNgJ/BiYU0Ff1lqdJWgF8m2TY7YMbSOdIqpFUU7t29Zb2Hj160KNHD4YMGQLASSedxFNPPbUz+c3MbBvKpdBsBk4GPinpO9vpewjwTN7yWOBjwP8Cv2hog4iYEhGDImJQRfuOW9o/+tGPst9++7FkyRIAHnzwQQ4++OCGdmFmZjupstQB6kTEWkkjgYclvRYR19XvI+lE4DjgwnrbbpT078BzkvpExDP1t92Wa665hrFjx7JhwwZ69uzJDTfcsKsfxczM8pRNoQGIiLckjQDmSnojbb5A0mnAnsBi4NiIeL2BbddJ+glwEYUNwQFQXV1NTU1NE6Q3M7OGlEWhiYgOee+XkwyFQXKDwGWNbDe03vJPMohnZma7oFyu0ZiZWQvV6gpN/+4dWTbpc6WOYWbWarS6QmNmZsXlQmNmZplyoTEzs0y50JiZWaZcaMzMLFMuNGZmlikXGjMzy5QLjZmZZcqFxszMMuVCY2ZmmXKhMTOzTLW6QrPo5dXb72RmZk2m1RUaMzMrrlZfaGpraznkkEMYOXJkqaOYmbVImRcaSbWScnmvidvo95CkQQ20nyHp9XTbpyX9TlL7dN03Jf1F0kJJD0o6YEfzXX311fTp02fHP5iZmRWkGGc06yKiOu81qX4HSRXb2cet6bZ9gQ3A6LR9PjAoIgYAvwOu2JFgK1asYMaMGZx11lk7spmZme2Akg2dSVom6T8l/Qn4ctp8mqQ/S1osaXAD21QCewJvA0TEnIhYm65+DOixIxnOP/98rrjiCtq0afUjiGZmmSnG/2H3qDd0Njpv3fqIOCoipqfLe0bEkcDXgOvz+o2WlANeBj4M/F8Dx5kA3NtQAEnnSKqRVFO7Nrnr7J577qFLly4cdthhu/bpzMysUaUYOrs1b92t9freAhARc4G9JXWq6xcR1cBHgUXAxfkbSToNGARc2VCAiJgSEYMiYlBF+44APPLII9x9991UVVVxyimnMHv2bE477bRd+6RmZvYBpR4zWlNvORpbjoggOZs5uq5N0meA7wKfj4j3Cz3w5ZdfzooVK1i2bBnTp0/n2GOP5eabb96x9GZmtl2lLjT1jQaQdBSwOiIa+nblUcBzab9DgF+TFJlVRUtpZmYFqyzCMfZIr6/UmRURDd7iDLwt6c/A3sD4vPbRafFpA6wAzkjbrwQ6ALdJAngpIj6/owGHDh3K0KFDd3QzMzMrgJLRqNajXbde8f7KpaWOYWbWrEh6MiI+8F3HQpTb0Fnm+nfvWOoIZmatSqsrNGZmVlwuNGZmlikXGjMzy5QLjZmZZcqFxszMMtXqbm+W9B6wpNQ5tqMz8EapQxTAOZuWczad5pARmk/O3YEVETFiZzZujYWmZmfvBS+W5pARnLOpOWfTaQ4ZofXk9NCZmZllyoXGzMwy1RoLzZRSByhAc8gIztnUnLPpNIeM0EpytrprNGZmVlyt8YzGzMyKyIXGzMwy1WoKjaQRkpZI+pukbT0Pp1hZrpe0StLivLYPS7pf0tL054fy1l2S5l4i6bNFyrifpDmSnpH0tKRvlGnO3SXNk7Qgzfm9csyZd+wKSfMl3VOuOSUtk7RIUk5STRnn7CTpd5KeTf87PaKccko6KP0d1r3elXR+OWXMO+4F6d+fxZJuSf9eNV3OiGjxL6CC5KmcPYG2wALg4BLmORo4FFic13YFMDF9PxH4Ufr+4DRvO+Bj6eeoKELGbsCh6fu9gL+mWcotp4AO6fvdgMeBw8stZ17ebwL/C9xTjn/u6bGXAZ3rtZVjzqnAWen7tkCncsyZHr8CeBU4oNwyAt2BF4A90uXfkjxcsslyFuWXXOoXcARwX97yJcAlJc5UxdaFZgnQLX3fDVjSUFbgPuCIEuS9CxhezjmB9sBTwJByzAn0AB4EjuUfhaYccy7jg4WmrHKSPIX3BdIbmso1Z97xjgMeKceMJIVmOfBhkqcu35PmbbKcrWXorO4XWWdF2lZOukbESoD0Z5e0veTZJVUBh5CcLZRdznQ4KgesAu6PiLLMCVwFfAvYnNdWjjkD+IOkJyWdU6Y5ewKvAzekQ5G/kbRnGeascwpwS/q+rDJGxMvAj4GXgJXA6oj4Q1PmbC2FRg20NZf7ukuaXVIH4PfA+RHxbmNdG2grSs6IqI2IapIzhsGS+jXSvSQ5JY0EVkXEk4Vu0kBbsf7cPxURhwLHA/8q6ehG+pYqZyXJ8PMvI+IQYA3J8M62lOz3Kakt8Hngtu11baCtGP9tfgj4Askw2L7AnpJOa2yTBtoazdlaCs0KYL+85R7AKyXKsi2vSeoGkP5clbaXLLuk3UiKzLSIuL1cc9aJiHeAh4ARlF/OTwGfl7QMmA4cK+nmMsxJRLyS/lwF3AEMLsOcK0gmeXw8Xf4dSeEpt5yQFOynIuK1dLncMn4GeCEiXo+IjcDtwJFNmbO1FJongF6SPpb+6+IU4O4SZ6rvbmBc+n4cyTWRuvZTJLWT9DGgFzAv6zCSBFwHPBMRPy3jnB+R1Cl9vwfJX5pnyy1nRFwSET0ioorkv7/ZEXFaueWUtKekverek4zVLy63nBHxKrBc0kFp0zDgL+WWMzWGfwyb1WUpp4wvAYdLap/+vR8GPNOkOYt1MazUL+AEkjunngO+W+Ist5CMhW4k+dfBBGAfkgvFS9OfH87r/9009xLg+CJlPIrkdHghkEtfJ5RhzgHA/DTnYuA/0/ayylkv81D+cTNAWeUkufaxIH09Xfd3pdxypsetBmrSP/s7gQ+VW06SG1TeBDrmtZVVxvS43yP5B9pi4CaSO8qaLKenoDEzs0y1lqEzMzMrERcaMzPLlAuNmZllyoXGzMwy5UJjZmaZcqGxZktSSPpJ3vJFki5ron3fKOmkptjXdo7z5XTm4TkF9v9OEx67k6SvNdX+zLbFhcaas/eBL0nqXOog+SRV7ED3CcDXIuLTBfZvsNAosaN/nzsBO1xodvDzmbnQWLO2ieRZ5hfUX1H/jETS39OfQyX9UdJvJf1V0iRJY5U802aRpAPzdvMZSQ+n/Uam21dIulLSE5IWSvpK3n7nSPpfYFEDecak+18s6Udp23+SfDH2V5KurNe/m6S5Sp5jsljSP0maBOyRtk2TVJWeDf03yazV+9V9znQfJ0m6MX3fVdIdSp7bs0DSkcAk4MB0f1emn+GevO0nSzojfb9M0n9K+hPwZUnHSXpU0lOSblMyJx7p7/Mv6e/mx4X9MVpLV1nqAGa76BfAQklX7MA2A4E+wFvA88BvImKwkoe7/RtwftqvCjgGOBCYI+njwL+QzG77SUntgEck/SHtPxjoFxEv5B9M0r7Aj4DDgLdJZkb+YkR8X9KxwEURUVMv46kkj7b4YXoG0T4iHpZ0XiQTiNbNqn0QcGZEfC1t29Zn/jnwx4gYle6vA8kklP3y9jd0O7+39RFxVHoGeTvwmYhYI+nbwDclTQZGAb0jIuqmBjJzobFmLSLelfQ/wNeBdQVu9kSk059Leg6oKxSLgPwhrN9GxGZgqaTngd4kc38NyDtb6kgy19MGYF79IpP6JPBQRLyeHnMaycPv7mwsI3C9kolN74yI3Db6vRgRjzX2YVPHkhRJIqIWWK28JyYW6Nb05+EkD796JC1sbYFHgXeB9cBvJM0gea6JmYfOrEW4iuRax555bZtI//tOJwpsm7fu/bz3m/OWN7P1P77qz88UJFOk/1tEVKevj0Xy7A5IpqpvyDZPM7YlIuaSFKOXgZsk/cs2utY/Zn7m3XfwsFt+Z9vYvu5YInnuT93v4OCImBARm0jO6n4PfBGYtYPHtxbKhcaavYh4i+TxsxPympeRDFVB8qyN3XZi11+W1Ca9btOTZALB+4Bz0zMNJH1CySzHjXkcOEZS53TYagzwx8Y2kHQAyfNrriWZRfvQdNXGumNvw2uS+qQ3BozKa38QODfdd4WkvYH3SB7TXedF4GAls/J2JJnFtyGPAZ9KhxJRMuvvJ9LrNB0jYibJ8GN1Y5/RWg8XGmspfgLk3312Lcn/3OeRPNp5W2cbjVlCUhDuBb4aEeuB35BMR/+UpMXAr9nOEHQ6THcJMIdkVuSnIuKuxrYhmeE5J2k+cCJwddo+heSa1LRtbDeRZMhqNskM4XW+AXxa0iLgSaBvRLxJMvy1WNKVEbGcpGAvBKaRzIrd0Od5neSZ8rdIWkhSeHqTFK170rY/0sBNGtY6efZmMzPLlM9ozMwsUy40ZmaWKRcaMzPLlAuNmZllyoXGzMwy5UJjZmaZcqExM7NM/X9NBGuncZxNrQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots()\n",
    "n_structures_per_kinase.plot(kind=\"barh\", ax=ax)\n",
    "ax.set_xlabel(\"Number of structures\")\n",
    "ax.set_ylabel(\"Kinase name\")\n",
    "for i, value in enumerate(n_structures_per_kinase):\n",
    "    ax.text(value, i, str(value), va=\"center\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that our dataset is highly imbalanced. While some kinases are structurally resolved very often, other kinases are not. We will have to keep this in mind when interpreting our results later."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare structures\n",
    "\n",
    "We will make a pairwise comparison of the structures' IFP using the Tanimoto/Jaccard distance as implemented in `sklearn.metrics.pairwise_distances`, which uses under the hood the method `scipy.spatial.distance`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prepare IFPs as `numpy` array\n",
    "\n",
    "KLIFS deposits the IFP as a string of 0's and 1's. We have to convert the IFPs to an array of boolean vectors (required by `scipy.spatial.distance` to be able to use the Jaccard distance). Each row in this array refers to one IFP, each columns to one of the IFP's features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "structure.klifs_id  kinase.klifs_name\n",
       "775                 EGFR                 0000000000000010000000000000000000000000000000...\n",
       "777                 EGFR                 0000000000000010000001000000000000000000000000...\n",
       "778                 EGFR                 0000000000000010000000000000000000000000000000...\n",
       "779                 EGFR                 0000000000000010000001000000000000000000000000...\n",
       "782                 EGFR                 0000000000000010000001000000000000000000000000...\n",
       "Name: interaction.fingerprint, dtype: string"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This is the KLIFS format of the IFP (structure KLIFS ID and kinase name set as index)\n",
    "ifp_series = structures_with_ifps_df.set_index([\"structure.klifs_id\", \"kinase.klifs_name\"])[\n",
    "    \"interaction.fingerprint\"\n",
    "]\n",
    "ifp_series.head()\n",
    "# NBVAL_CHECK_OUTPUT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "structure.klifs_id  kinase.klifs_name\n",
       "775                 EGFR                 [False, False, False, False, False, False, Fal...\n",
       "777                 EGFR                 [False, False, False, False, False, False, Fal...\n",
       "778                 EGFR                 [False, False, False, False, False, False, Fal...\n",
       "779                 EGFR                 [False, False, False, False, False, False, Fal...\n",
       "782                 EGFR                 [False, False, False, False, False, False, Fal...\n",
       "Name: interaction.fingerprint, dtype: object"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cast \"0\" and \"1\" to boolean False and True\n",
    "ifp_series = ifp_series.apply(lambda x: [True if i == \"1\" else False for i in x])\n",
    "ifp_series.head()\n",
    "# NBVAL_CHECK_OUTPUT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[False, False, False, ..., False, False, False],\n",
       "       [False, False, False, ..., False, False, False],\n",
       "       [False, False, False, ..., False, False, False],\n",
       "       ...,\n",
       "       [False, False, False, ..., False, False, False],\n",
       "       [False, False, False, ..., False, False, False],\n",
       "       [False, False, False, ..., False, False, False]])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert to numpy array\n",
    "ifps_array = np.array(ifp_series.to_list())\n",
    "ifps_array\n",
    "# NBVAL_CHECK_OUTPUT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calculate pairwise Jaccard distances\n",
    "\n",
    "The Jaccard distance, defined below, is often used in case of binary fingerprints: \n",
    "\n",
    "$$\n",
    "d_J(A,B) = 1 - J(A,B) = \\frac{\\mid A \\cup B \\mid - \\mid A \\cap B \\mid}{\\mid A \\cup B \\mid}.\n",
    "$$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "structure_distance_matrix_array = pairwise_distances(ifps_array, metric=\"jaccard\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Structure distance matrix size: (1466, 1466)\n",
      "Show matrix subset:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>structure.klifs_id</th>\n",
       "      <th>775</th>\n",
       "      <th>777</th>\n",
       "      <th>778</th>\n",
       "      <th>779</th>\n",
       "      <th>782</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure.klifs_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>775</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.720000</td>\n",
       "      <td>0.542857</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.483871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777</th>\n",
       "      <td>0.720000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.741935</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.703704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778</th>\n",
       "      <td>0.542857</td>\n",
       "      <td>0.741935</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.696970</td>\n",
       "      <td>0.411765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>779</th>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.696970</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.655172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>782</th>\n",
       "      <td>0.483871</td>\n",
       "      <td>0.703704</td>\n",
       "      <td>0.411765</td>\n",
       "      <td>0.655172</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "structure.klifs_id       775       777       778       779       782\n",
       "structure.klifs_id                                                  \n",
       "775                 0.000000  0.720000  0.542857  0.500000  0.483871\n",
       "777                 0.720000  0.000000  0.741935  0.500000  0.703704\n",
       "778                 0.542857  0.741935  0.000000  0.696970  0.411765\n",
       "779                 0.500000  0.500000  0.696970  0.000000  0.655172\n",
       "782                 0.483871  0.703704  0.411765  0.655172  0.000000"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create DataFrame with structure KLIFS IDs as index/columns\n",
    "structure_klifs_ids = ifp_series.index.get_level_values(0)\n",
    "structure_distance_matrix_df = pd.DataFrame(\n",
    "    structure_distance_matrix_array, index=structure_klifs_ids, columns=structure_klifs_ids\n",
    ")\n",
    "print(f\"Structure distance matrix size: {structure_distance_matrix_df.shape}\")\n",
    "print(\"Show matrix subset:\")\n",
    "structure_distance_matrix_df.iloc[:5, :5]\n",
    "# NBVAL_CHECK_OUTPUT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Map structure to kinase distance matrix\n",
    "\n",
    "Note: So far we compared individual structures, but we want to compare kinases (which can be represented by several structures, as plotted above).\n",
    "\n",
    "First, as an intermediate step, we will create a structure distance matrix but &mdash; instead of labeling the data with structure KLIFS IDs &mdash; we use the corresponding kinase name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Show matrix subset:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>kinase.klifs_name</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>EGFR</th>\n",
       "      <th>EGFR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kinase.klifs_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EGFR</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.720000</td>\n",
       "      <td>0.542857</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.483871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGFR</th>\n",
       "      <td>0.720000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.741935</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.703704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGFR</th>\n",
       "      <td>0.542857</td>\n",
       "      <td>0.741935</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.696970</td>\n",
       "      <td>0.411765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGFR</th>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.696970</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.655172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGFR</th>\n",
       "      <td>0.483871</td>\n",
       "      <td>0.703704</td>\n",
       "      <td>0.411765</td>\n",
       "      <td>0.655172</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "kinase.klifs_name      EGFR      EGFR      EGFR      EGFR      EGFR\n",
       "kinase.klifs_name                                                  \n",
       "EGFR               0.000000  0.720000  0.542857  0.500000  0.483871\n",
       "EGFR               0.720000  0.000000  0.741935  0.500000  0.703704\n",
       "EGFR               0.542857  0.741935  0.000000  0.696970  0.411765\n",
       "EGFR               0.500000  0.500000  0.696970  0.000000  0.655172\n",
       "EGFR               0.483871  0.703704  0.411765  0.655172  0.000000"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Copy distance matrix to kinase matrix\n",
    "kinase_distance_matrix_df = structure_distance_matrix_df.copy()\n",
    "# Replace structure KLIFS IDs with the structures' kinase names\n",
    "kinase_names = ifp_series.index.get_level_values(1)\n",
    "kinase_distance_matrix_df.index = kinase_names\n",
    "kinase_distance_matrix_df.columns = kinase_names\n",
    "print(\"Show matrix subset:\")\n",
    "kinase_distance_matrix_df.iloc[:5, :5]\n",
    "# NBVAL_CHECK_OUTPUT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A kinase pair can be represented by many different structure pairs which are associated with different distance values. \n",
    "\n",
    "For example (in the demo mode), if we compare all EGFR structure to each other, the range of distance values is already quite high because we can observe different binding modes of the co-crystallized ligands."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "tags": [
     "nbsphinx-thumbnail"
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAD4CAYAAADhNOGaAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAAUDUlEQVR4nO3dfbRldX3f8fdnBikDFR1gEg0DzOhCyNCAyhWNmtQ2hIEsV4jRpKLVig8EA0aaPkCMK2LT1doudUkVM45Ggq6uUi0aMSUyDSuiC+JyLuFxUHAcCoyQZEAKBFQy8O0f51w9XO7DvjN37zP37vdrrbPm7MfzPX/M+dy9f7/9+6WqkCT114pxFyBJGi+DQJJ6ziCQpJ4zCCSp5wwCSeq5/cZdwEIddthhtW7dunGXIUlLyvXXX39/Va2ZaduSC4J169YxOTk57jIkaUlJctds27w1JEk9ZxBIUs8ZBJLUcwaBJPWcQSBJPWcQSItg48aNrFixgiSsWLGCjRs3jrskqTGDQNpLGzduZMuWLUyN5FtVbNmyxTDQkmEQSHtpy5YtC1ov7WsMAknqOYNAknrOIJAWyXHHHcddd93FcccdN+5SpAVZcmMNSfuqbdu2cdRRR427DGnBvCKQpJ4zCCSp5wwCaS+dcsopC1ov7WtsI5BmkWSvjt+yZUujc0w9iCaNi0EgzWJPfqCT+MOuJcdbQ5LUcwaBJPWcQSBJPWcQSFLPGQSS1HMGgST1nEEgST1nEEhSzxkEktRzBoEk9ZxBIEk9ZxBIUs8ZBJLUcwaBJPWcQSBJPWcQSFLPGQSS1HMGgST1nEEgST3XahAkOTXJ7Um2J7lghu3PSvLlJDcl2ZbkzDbrkSQ93YKCIMnqJMc33HclcDFwGrABOCPJhmm7nQPcVlUnAK8CPpRk/4XUJEnaO/MGQZKvJjk4ySHATcAlST7c4NwnAdurakdVPQ5cBpw+bZ8CnpkkwD8Gvg/sXtA3kCTtlSZXBM+qqoeBXwcuqaoTgZMbHHc4cM/I8s7hulEfA34WuBe4BXh3VT05/URJzkoymWRy165dDT5aktRUkyDYL8lzgd8E/mwB584M62ra8kbgRuBngBcCH0ty8NMOqtpcVRNVNbFmzZoFlCBJmk+TIPgPwFXAd6tqa5LnAd9pcNxO4IiR5bUM/vIfdSbwhRrYDtwJHNvg3JKkRTJvEFTV56vq+Kp653B5R1W9tsG5twJHJ1k/bAB+PXDFtH3uBn4JIMlPA8cAOxbyBSRJe6dJY/ELklyd5Nbh8vFJ3jvfcVW1GziXwdXEt4DPVdW2JGcnOXu42x8CL09yC3A1cH5V3b+nX0aStHCpmn7bftoOyTXAvwM+UVUvGq67tar+SQf1Pc3ExERNTk6O46OleSVhvv9T0jgkub6qJmba1qSN4MCq+ua0dXbxlKRlokkQ3J/k+Qx7/CR5HXBfq1VJkjqzX4N9zgE2A8cm+R6Dnj1vbLUqSVJn5gyC4TAR76yqk5McBKyoqke6KU2S1IU5g6Cqnkhy4vD9o92UJEnqUpNbQzckuQL4PPDjMKiqL7RWlSSpM02C4BDgAeCfj6wrwCCQpGVg3iCoKucIkKRlbN4gSHIJTx8sjqp6aysVSZI61eTW0OiIowcAr+Hpg8dJkpaoJreGLh9dTvI/gL9orSJJUqf2ZM7io4EjF7sQSdJ4NGkjeISnthH8DXB+axVJkjrV5NbQM7soRJI0Hk3mI7i6yTpJ0tI06xVBkgOAA4HDkqzmJ3MQH8xgjmFJ0jIw162h3wLOY/Cjfz0/CYKHgYvbLUuS1JVZg6CqLgIuSvKuqvpohzVJkjrUpPvok0mePbWQZHWS326vJElSl5oEwTuq6v9NLVTVg8A7WqtIktSpJkGwIslU+8DUZDX7t1eSJKlLTcYaugr4XJJNDB4sOxv4SqtVSZI60yQIzmfQg+idDHoObQE+1WZRkqTuNHmy+Engj4YvSdIy02SsoTuZeT6C57VSkSSpU01uDU2MvD8A+A0G01dKkpaBeXsNVdUDI6/vVdVHeOr8xZKkJazJraEXjyyuYHCF4IikkrRMNLk19KGR97uBO4HfbKccSVLX5gyCJCuATVX1PzuqR2rFIYccwoMPPtjJZ408f9mK1atX8/3vf7/Vz1C/zBkEVfVkknMAg0BL2oMPPkjV0zq/LUltB436p8kQE/8nyb9NckSSQ6ZerVcmSepEkzaCtw7/PWdkXQE+RyBJy0CTK4Kfrar1oy9gQ5OTJzk1ye1Jtie5YJZ9XpXkxiTbklyzkOIlSXuvSRBc13DdUwxHKb0YOI1BcJyRZMO0fZ4NfBz41ao6jsHDapKkDs01Z/FzgMOBVUlexFPnLD6wwblPArZX1Y7h+S4DTgduG9nnDcAXqupugKr6uwV/A0nSXpmrjWAj8BZgLYNnCaaC4BHgPQ3OfThwz8jyTuCl0/Z5AfCMJF9l8JDaRVX1meknSnIWcBbAkUce2eCjJUlNzTVn8aXApUleW1WX78G5Z+rjNr3/3n7AicAvAauAv0ryjaq6Y1otm4HNABMTE8ujD6Ak7SOatBGsTXJwBj6V5K+TnNLguJ3AEaPnAe6dYZ+vVNWjVXU/8DXghEaVS5IWRZMgeGtVPQycAvwUcCbwgQbHbQWOTrI+yf7A64Erpu3zJeAXkuyX5EAGt46+1bh6SdJea/IcwdQtnl8BLqmqm0bnMJ5NVe1Oci6DqS5XAp+uqm1Jzh5u31RV30ryFeBm4EngU1V16x59E0nSHsl8j90nuYRBw+96BrdtVgJfraoT2y/v6SYmJmpycnIcH60lLMmyGmJiuXwXdSfJ9VU1MdO2JlcEbwNeCOyoqseSHMrg9pAkaRloOmfxX48sPwA80GZRkqTuNGksliQtYwaBJPVckzaCqXGDfnp0/6lhISRJS1uTOYvfBbwP+FsGXTxh8ITw8S3WJUnqSJMrgncDxwwbiSVJy0yTILgHeKjtQqQ21fsOhgufNe4yFkW97+Bxl6BlpkkQ7AC+muR/Az+aWllVH26tKmmR5f0PL5uHsJJQF467Ci0nTYLg7uFr/+FLkrSMNHmg7P0ASQ6qqkfbL0mS1KV5nyNI8vNJbmM4KmiSE5J8vPXKJEmdaPJA2UcYzFb2AEBV3QT8Yos1SZI61OjJ4qq6Z9qqJ1qoRZI0Bo26jyZ5OVDDCWZ+ByePkaRlo8kVwdnAOQzmJNjJYEjqc1qsSZLUoSa9hu4H3thBLZKkMWjSa+gFSa5Ocutw+fgk722/NElSF5rcGvok8HvAPwBU1c0MJqKXJC0DTYLgwKr65rR1u9soRpLUvSZBcH+S5zMYepokrwPua7UqSVJnmnQfPQfYDByb5HvAndh4LEnLRpNeQzuAk5McBKyoqkfaL0uS1JUmvYa+m+S/A28Cjmi/JElSl5q0EWwAPgEcCnwwyY4kX2y3LElSV5oEwRMMuo4+wWDO4r8F/q7NoiRJ3WnSWPwwcAvwYeCTzl2spSrJuEtYFKtXrx53CVpmmgTBGcArgd8G3p7kOuBrVXV1q5VJi6iraSqTLJspMdUfTXoNfQn4UpJjgdOA84B/D6xqtzRJUhea9Bq6PMl3gYuAg4A3A16bStIy0eTW0EXAtVX148lokvyj9kqSJHWp0VSVoyEw9FdtFCNJ6t6sQZDkOUlOBFYleVGSFw9frwIObHLyJKcmuT3J9iQXzLHfS5I8MRzHSJLUobluDW0E3gKsZdB1dMrDwHvmO3GSlcDFwC8zmNlsa5Irquq2Gfb7L8BVC6pckrQoZg2CqroUuDTJa6vq8j0490nA9uFYRSS5DDgduG3afu8CLgdesgefIUnaS03aCK5N8sdJ/hwgyYYkb2tw3OHAPSPLO4frfizJ4cBrgE1znSjJWUkmk0zu2rWrwUdLkppqEgSXMLht8zPD5TsYPEswn5ke45z+pM1HgPNnaIx+6kFVm6tqoqom1qxZ0+CjJUlNNQmCw6rqcwzGGaKqdjMYd2g+O3nqaKVrgXun7TMBXJbk/wKvAz6e5NcanFuStEiaPEfwaJJD+ckMZS8DHmpw3Fbg6CTrge8xmOf4DaM7VNX6qfdJ/gT4s6r600aVS5IWRZMg+F3gCuD5Sa4F1jD4631OVbU7ybkMbiutBD5dVduSnD3cPme7gCSpG2kyQFaS/YBjGNz3v72q/qHtwmYzMTFRk5OT4/p4aU4OOqd9VZLrq2pipm3zXhEkOYDByKOvZHB76OtJNlXVDxe3TEnSODS5NfQZ4BHgo8PlM4DPAr/RVlGSpO40CYJjquqEkeW/THJTWwVJkrrVpPvoDcOeQgAkeSlwbXslSZK6NOsVQZJbGLQJPAN4c5K7h8tH8fRhIiRJS9Rct4Ze3VkVkqSxmWvQubu6LESSNB5N2ggkScuYQSBJPdcoCJIcleTk4ftVSZ7ZblmSpK7MGwRJ3gH8L+ATw1VrgT9tsSZJUoeaXBGcA7yCwRSVVNV3gJ9qsyhJUneaBMGPqurxqYXhAHSOqiVJy0STILgmyXuAVUl+Gfg88OV2y5IkdaVJEFwA7AJuAX4LuBJ4b5tFSZK602TQuVUMJpX5JECSlcN1j7VZmCSpG02uCK5m8MM/ZRXwF+2UI0nqWpMgOKCq/n5qYfj+wPZKkiR1qUkQPJrkxVMLSU4EftBeSZKkLjVpIzgP+HySe4fLzwX+RWsVSZI6NW8QVNXWJMfyk8nrvz3OyeslSYuryRUBwEuAdcP9X5SEqvpMa1VJkjozbxAk+SzwfOBG4Inh6mIwqb0kaYlrckUwAWyoKoeVkKRlqEmvoVuB57RdiCRpPJpcERwG3Jbkm8CPplZW1a+2VpUkqTNNguDCtouQJI1Pk+6j13RRiCRpPJrMUPayJFuT/H2Sx5M8keThLoqTJLWvSWPxx4AzgO8wGHDu7cN1kqRloNEDZVW1PcnKqnoCuCTJdS3XJUnqSJMgeCzJ/sCNSf4rcB9wULtlSZK60uTW0JuG+50LPAocAfx6m0VJkrrTJAh+rap+WFUPV9X7q+p3gVc3OXmSU5PcnmR7kgtm2P7GJDcPX9clOWGhX0CStHeaBMG/mmHdW+Y7aDil5cXAacAG4IwkG6btdifwT6vqeOAPgc0N6pEkLaJZ2wiSnAG8AVif5IqRTQcDDzQ490nA9qraMTzfZcDpwG1TO1TVaKPzN4C1zUuXJC2GuRqLr2PQMHwY8KGR9Y8ANzc49+HAPSPLO4GXzrH/24A/n2lDkrOAswCOPPLIBh8tSWpq1iCoqruAu5KcDPygqp5M8gLgWOCWBufOTKedccfknzEIglfOUstmhreNJiYmHAVVkhZRkzaCrwEHJDkcuBo4E/iTBsftZNDDaMpa4N7pOyU5HvgUcHpVNbnlJElaRE2CIFX1GIMuox+tqtcwaPydz1bg6CTrh88hvB4YbWsgyZHAF4A3VdUdCytdkrQYmjxQliQ/D7yRwe2bRsdV1e4k5wJXASuBT1fVtiRnD7dvAv4AOBT4eBKA3VU1sfCvIUnaU02C4Dzg94AvDn/Inwf8ZZOTV9WVwJXT1m0aef92BmMXSZLGpOkw1NeMLO8AfqfNoqSlZnhF+5T3zu6qpWKu5wg+UlXnJfkyM/T2cYYyaWA0BKavNwy0FMx1RfDZ4b8f7KIQaV8z2w/8Yp/DsNC4zfUcwfXDf69Jsmb4fldXhUnj1vQHeq4fe3/ktRTM2n00AxcmuR/4NnBHkl1J/qC78iRJbZvrOYLzgFcAL6mqQ6tqNYMhIl6R5F93UZwkqX1zBcGbgTOq6s6pFcMeQ/9yuE2StAzMFQTPqKr7p68cthM8o72SJEldmisIHt/DbZKkJWSu7qMnJHl4hvUBDmipHklSx+bqPrqyy0IkSePRZPRRSdIyZhBIUs8ZBJLUcwaBJPWcQSBJPWcQSFLPGQSS1HMGgST1nEEgST1nEEhSzxkEktRzBoEk9ZxBIEk9ZxBIUs8ZBJLUcwaBJPWcQSBJPWcQSFLPGQSS1HMGgST1nEEgST1nEEhSz7UaBElOTXJ7ku1JLphhe5L8t+H2m5O8uM16pDZU1YLWS/ua1oIgyUrgYuA0YANwRpIN03Y7DTh6+DoL+KO26pHakmRB66V9TZtXBCcB26tqR1U9DlwGnD5tn9OBz9TAN4BnJ3luizVJramqH7+kpaTNIDgcuGdkeedw3UL3IclZSSaTTO7atWvRC5WkPmszCGa6Lp7+p1KTfaiqzVU1UVUTa9asWZTiJEkD+7V47p3AESPLa4F792AfaUmwTUBLVZtXBFuBo5OsT7I/8Hrgimn7XAG8edh76GXAQ1V1X4s1SYvOXkNa6lq7Iqiq3UnOBa4CVgKfrqptSc4ebt8EXAn8CrAdeAw4s616pDb5o6+lrM1bQ1TVlQx+7EfXbRp5X8A5bdYgSZqbTxZLUs8ZBJLUcwaBJPWcQSBJPZel1tshyS7grnHXIc3iMOD+cRchzeCoqprxidwlFwTSvizJZFVNjLsOaSG8NSRJPWcQSFLPGQTS4to87gKkhbKNQJJ6zisCSeo5g0CSes4gkGaR5IkkN468Lhiu3y/Jf0rynZFtvz/HceuSvCrJQ0luSPLtJB8c3zeTnqrV0UelJe4HVfXCGdb/R+A5wM9V1Q+TPBP4N3Mdl2Qd8PWqenWSVcANSb5YVde2U7rUnEEgLUCSA4F3AOuq6ocAVfUIcGHTc1TVD5LcyAzzc0vjYBBIs1s1/MGe8p+BbwF3D3/8mxx3Z1W9ZnRjktXA0cDXFrFWaY8ZBNLsZrrFc/y05TOBdwOHAi+vqntmOm7oF5LcDBwDfKCq/qaVqqUFsrFYWpjtwJHDdgGq6pLhj/5DDKZkncvXq+p44OeAdyZ5YZuFSk0ZBNICVNVjwB8DH0tyAECSlcD+CzjHHQxuM53fSpHSAhkE0uxWTesG+oHh+t8H7gNuTXID8HXgUuDeBZx7E/CLSdYvbsnSwjnEhCT1nFcEktRzBoEk9ZxBIEk9ZxBIUs8ZBJLUcwaBJPWcQSBJPff/ASOIW9hOGlg1AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "if DEMO:\n",
    "    example = \"EGFR\"\n",
    "else:\n",
    "    example = kinase_selection_df[\"kinase_klifs\"][0]\n",
    "\n",
    "# Select EGFR-EGFR structure pairs only\n",
    "D = kinase_distance_matrix_df.loc[example, example]\n",
    "# Extract all pairwise distances without identical structure pairs\n",
    "# = lower triangular matrix without the diagonal\n",
    "D_condensed = distance.squareform(D)\n",
    "# Plot pairwise distances\n",
    "plt.boxplot(D_condensed)\n",
    "plt.xticks([1], [example])\n",
    "plt.ylabel(\"Distance between structures\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this talktorial, we will consider per kinase pair the two structures that show the most similar binding mode for their co-crystallized ligands. Hence, we select the structure pair with the minimum IFP distance as representative for a kinase pair."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We unstack the matrix (each pairwise comparison in a single row)\n",
    "# We group by kinase names (level=[0, 1] ensures that the order of the kinases is ignored\n",
    "# We take the minimum value in each kinase pair group\n",
    "# We unstack the remaining data points\n",
    "kinase_distance_matrix_df = (\n",
    "    kinase_distance_matrix_df.unstack().groupby(level=[0, 1]).min().unstack(level=1)\n",
    ")\n",
    "kinase_distance_matrix_df.index.name = None\n",
    "kinase_distance_matrix_df.columns.name = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Structure matrix of shape (1466, 1466) reduced to kinase matrix of shape (9, 9).\n"
     ]
    }
   ],
   "source": [
    "print(\n",
    "    f\"Structure matrix of shape {structure_distance_matrix_df.shape} \"\n",
    "    f\"reduced to kinase matrix of shape {kinase_distance_matrix_df.shape}.\"\n",
    ")\n",
    "# NBVAL_CHECK_OUTPUT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_dd0ab_row0_col0, #T_dd0ab_row1_col1, #T_dd0ab_row2_col2, #T_dd0ab_row3_col3, #T_dd0ab_row4_col4, #T_dd0ab_row5_col5, #T_dd0ab_row6_col6, #T_dd0ab_row7_col7, #T_dd0ab_row8_col8 {\n",
       "  background-color: #ebf3eb;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row0_col1, #T_dd0ab_row2_col7, #T_dd0ab_row8_col0 {\n",
       "  background-color: #4ea64e;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row0_col2, #T_dd0ab_row4_col0, #T_dd0ab_row5_col3 {\n",
       "  background-color: #3f9f3f;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row0_col3 {\n",
       "  background-color: #49a449;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row0_col4 {\n",
       "  background-color: #4ba54a;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row0_col5, #T_dd0ab_row3_col4 {\n",
       "  background-color: #379b37;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row0_col6 {\n",
       "  background-color: #209020;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row0_col7 {\n",
       "  background-color: #1d8e1d;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row0_col8 {\n",
       "  background-color: #45a245;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row1_col0 {\n",
       "  background-color: #5eae5e;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row1_col2 {\n",
       "  background-color: #a6d2a6;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row1_col3 {\n",
       "  background-color: #319831;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row1_col4, #T_dd0ab_row2_col4 {\n",
       "  background-color: #82c082;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row1_col5, #T_dd0ab_row8_col2, #T_dd0ab_row8_col5 {\n",
       "  background-color: #80bf80;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row1_col6 {\n",
       "  background-color: #a2cfa2;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row1_col7 {\n",
       "  background-color: #5cad5c;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row1_col8 {\n",
       "  background-color: #61af61;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row2_col0 {\n",
       "  background-color: #68b368;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row2_col1 {\n",
       "  background-color: #b2d7b1;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row2_col3, #T_dd0ab_row8_col3 {\n",
       "  background-color: #81bf81;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row2_col5 {\n",
       "  background-color: #8ec58e;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row2_col6 {\n",
       "  background-color: #6eb66e;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row2_col8 {\n",
       "  background-color: #95c995;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row3_col0 {\n",
       "  background-color: #329832;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row3_col1, #T_dd0ab_row3_col5, #T_dd0ab_row3_col7, #T_dd0ab_row7_col0, #T_dd0ab_row7_col2, #T_dd0ab_row7_col3, #T_dd0ab_row7_col4, #T_dd0ab_row7_col6, #T_dd0ab_row7_col8 {\n",
       "  background-color: #008000;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row3_col2 {\n",
       "  background-color: #4ba54b;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row3_col6, #T_dd0ab_row7_col5 {\n",
       "  background-color: #198c19;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row3_col8 {\n",
       "  background-color: #6bb46b;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row4_col1 {\n",
       "  background-color: #6fb76f;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row4_col2, #T_dd0ab_row6_col4 {\n",
       "  background-color: #57ab57;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row4_col3 {\n",
       "  background-color: #43a143;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row4_col5 {\n",
       "  background-color: #cae3ca;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row4_col6 {\n",
       "  background-color: #349934;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row4_col7 {\n",
       "  background-color: #108810;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row4_col8 {\n",
       "  background-color: #d5e9d5;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row5_col0 {\n",
       "  background-color: #55a955;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row5_col1 {\n",
       "  background-color: #88c388;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row5_col2 {\n",
       "  background-color: #84c084;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row5_col4 {\n",
       "  background-color: #d1e6d1;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row5_col6 {\n",
       "  background-color: #61b061;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row5_col7 {\n",
       "  background-color: #50a750;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row5_col8 {\n",
       "  background-color: #8dc58d;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row6_col0 {\n",
       "  background-color: #3e9e3e;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row6_col1 {\n",
       "  background-color: #a6d1a6;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row6_col2, #T_dd0ab_row6_col5 {\n",
       "  background-color: #5fae5f;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row6_col3 {\n",
       "  background-color: #4da64d;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row6_col7 {\n",
       "  background-color: #3a9c3a;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row6_col8 {\n",
       "  background-color: #3c9d3c;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row7_col1 {\n",
       "  background-color: #359a35;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row8_col1 {\n",
       "  background-color: #5aac5a;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row8_col4 {\n",
       "  background-color: #d8ead8;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_dd0ab_row8_col6 {\n",
       "  background-color: #299429;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_dd0ab_row8_col7 {\n",
       "  background-color: #289328;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_dd0ab\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_dd0ab_level0_col0\" class=\"col_heading level0 col0\" >BRAF</th>\n",
       "      <th id=\"T_dd0ab_level0_col1\" class=\"col_heading level0 col1\" >CDK2</th>\n",
       "      <th id=\"T_dd0ab_level0_col2\" class=\"col_heading level0 col2\" >EGFR</th>\n",
       "      <th id=\"T_dd0ab_level0_col3\" class=\"col_heading level0 col3\" >ErbB2</th>\n",
       "      <th id=\"T_dd0ab_level0_col4\" class=\"col_heading level0 col4\" >KDR</th>\n",
       "      <th id=\"T_dd0ab_level0_col5\" class=\"col_heading level0 col5\" >LCK</th>\n",
       "      <th id=\"T_dd0ab_level0_col6\" class=\"col_heading level0 col6\" >MET</th>\n",
       "      <th id=\"T_dd0ab_level0_col7\" class=\"col_heading level0 col7\" >p110a</th>\n",
       "      <th id=\"T_dd0ab_level0_col8\" class=\"col_heading level0 col8\" >p38a</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_dd0ab_level0_row0\" class=\"row_heading level0 row0\" >BRAF</th>\n",
       "      <td id=\"T_dd0ab_row0_col0\" class=\"data row0 col0\" >0.000</td>\n",
       "      <td id=\"T_dd0ab_row0_col1\" class=\"data row0 col1\" >0.300</td>\n",
       "      <td id=\"T_dd0ab_row0_col2\" class=\"data row0 col2\" >0.278</td>\n",
       "      <td id=\"T_dd0ab_row0_col3\" class=\"data row0 col3\" >0.394</td>\n",
       "      <td id=\"T_dd0ab_row0_col4\" class=\"data row0 col4\" >0.364</td>\n",
       "      <td id=\"T_dd0ab_row0_col5\" class=\"data row0 col5\" >0.320</td>\n",
       "      <td id=\"T_dd0ab_row0_col6\" class=\"data row0 col6\" >0.368</td>\n",
       "      <td id=\"T_dd0ab_row0_col7\" class=\"data row0 col7\" >0.500</td>\n",
       "      <td id=\"T_dd0ab_row0_col8\" class=\"data row0 col8\" >0.333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_dd0ab_level0_row1\" class=\"row_heading level0 row1\" >CDK2</th>\n",
       "      <td id=\"T_dd0ab_row1_col0\" class=\"data row1 col0\" >0.300</td>\n",
       "      <td id=\"T_dd0ab_row1_col1\" class=\"data row1 col1\" >0.000</td>\n",
       "      <td id=\"T_dd0ab_row1_col2\" class=\"data row1 col2\" >0.111</td>\n",
       "      <td id=\"T_dd0ab_row1_col3\" class=\"data row1 col3\" >0.452</td>\n",
       "      <td id=\"T_dd0ab_row1_col4\" class=\"data row1 col4\" >0.238</td>\n",
       "      <td id=\"T_dd0ab_row1_col5\" class=\"data row1 col5\" >0.190</td>\n",
       "      <td id=\"T_dd0ab_row1_col6\" class=\"data row1 col6\" >0.133</td>\n",
       "      <td id=\"T_dd0ab_row1_col7\" class=\"data row1 col7\" >0.348</td>\n",
       "      <td id=\"T_dd0ab_row1_col8\" class=\"data row1 col8\" >0.278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_dd0ab_level0_row2\" class=\"row_heading level0 row2\" >EGFR</th>\n",
       "      <td id=\"T_dd0ab_row2_col0\" class=\"data row2 col0\" >0.278</td>\n",
       "      <td id=\"T_dd0ab_row2_col1\" class=\"data row2 col1\" >0.111</td>\n",
       "      <td id=\"T_dd0ab_row2_col2\" class=\"data row2 col2\" >0.000</td>\n",
       "      <td id=\"T_dd0ab_row2_col3\" class=\"data row2 col3\" >0.258</td>\n",
       "      <td id=\"T_dd0ab_row2_col4\" class=\"data row2 col4\" >0.238</td>\n",
       "      <td id=\"T_dd0ab_row2_col5\" class=\"data row2 col5\" >0.167</td>\n",
       "      <td id=\"T_dd0ab_row2_col6\" class=\"data row2 col6\" >0.227</td>\n",
       "      <td id=\"T_dd0ab_row2_col7\" class=\"data row2 col7\" >0.381</td>\n",
       "      <td id=\"T_dd0ab_row2_col8\" class=\"data row2 col8\" >0.174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_dd0ab_level0_row3\" class=\"row_heading level0 row3\" >ErbB2</th>\n",
       "      <td id=\"T_dd0ab_row3_col0\" class=\"data row3 col0\" >0.394</td>\n",
       "      <td id=\"T_dd0ab_row3_col1\" class=\"data row3 col1\" >0.452</td>\n",
       "      <td id=\"T_dd0ab_row3_col2\" class=\"data row3 col2\" >0.258</td>\n",
       "      <td id=\"T_dd0ab_row3_col3\" class=\"data row3 col3\" >0.000</td>\n",
       "      <td id=\"T_dd0ab_row3_col4\" class=\"data row3 col4\" >0.406</td>\n",
       "      <td id=\"T_dd0ab_row3_col5\" class=\"data row3 col5\" >0.419</td>\n",
       "      <td id=\"T_dd0ab_row3_col6\" class=\"data row3 col6\" >0.382</td>\n",
       "      <td id=\"T_dd0ab_row3_col7\" class=\"data row3 col7\" >0.571</td>\n",
       "      <td id=\"T_dd0ab_row3_col8\" class=\"data row3 col8\" >0.258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_dd0ab_level0_row4\" class=\"row_heading level0 row4\" >KDR</th>\n",
       "      <td id=\"T_dd0ab_row4_col0\" class=\"data row4 col0\" >0.364</td>\n",
       "      <td id=\"T_dd0ab_row4_col1\" class=\"data row4 col1\" >0.238</td>\n",
       "      <td id=\"T_dd0ab_row4_col2\" class=\"data row4 col2\" >0.238</td>\n",
       "      <td id=\"T_dd0ab_row4_col3\" class=\"data row4 col3\" >0.406</td>\n",
       "      <td id=\"T_dd0ab_row4_col4\" class=\"data row4 col4\" >0.000</td>\n",
       "      <td id=\"T_dd0ab_row4_col5\" class=\"data row4 col5\" >0.059</td>\n",
       "      <td id=\"T_dd0ab_row4_col6\" class=\"data row4 col6\" >0.333</td>\n",
       "      <td id=\"T_dd0ab_row4_col7\" class=\"data row4 col7\" >0.533</td>\n",
       "      <td id=\"T_dd0ab_row4_col8\" class=\"data row4 col8\" >0.043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_dd0ab_level0_row5\" class=\"row_heading level0 row5\" >LCK</th>\n",
       "      <td id=\"T_dd0ab_row5_col0\" class=\"data row5 col0\" >0.320</td>\n",
       "      <td id=\"T_dd0ab_row5_col1\" class=\"data row5 col1\" >0.190</td>\n",
       "      <td id=\"T_dd0ab_row5_col2\" class=\"data row5 col2\" >0.167</td>\n",
       "      <td id=\"T_dd0ab_row5_col3\" class=\"data row5 col3\" >0.419</td>\n",
       "      <td id=\"T_dd0ab_row5_col4\" class=\"data row5 col4\" >0.059</td>\n",
       "      <td id=\"T_dd0ab_row5_col5\" class=\"data row5 col5\" >0.000</td>\n",
       "      <td id=\"T_dd0ab_row5_col6\" class=\"data row5 col6\" >0.250</td>\n",
       "      <td id=\"T_dd0ab_row5_col7\" class=\"data row5 col7\" >0.375</td>\n",
       "      <td id=\"T_dd0ab_row5_col8\" class=\"data row5 col8\" >0.190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_dd0ab_level0_row6\" class=\"row_heading level0 row6\" >MET</th>\n",
       "      <td id=\"T_dd0ab_row6_col0\" class=\"data row6 col0\" >0.368</td>\n",
       "      <td id=\"T_dd0ab_row6_col1\" class=\"data row6 col1\" >0.133</td>\n",
       "      <td id=\"T_dd0ab_row6_col2\" class=\"data row6 col2\" >0.227</td>\n",
       "      <td id=\"T_dd0ab_row6_col3\" class=\"data row6 col3\" >0.382</td>\n",
       "      <td id=\"T_dd0ab_row6_col4\" class=\"data row6 col4\" >0.333</td>\n",
       "      <td id=\"T_dd0ab_row6_col5\" class=\"data row6 col5\" >0.250</td>\n",
       "      <td id=\"T_dd0ab_row6_col6\" class=\"data row6 col6\" >0.000</td>\n",
       "      <td id=\"T_dd0ab_row6_col7\" class=\"data row6 col7\" >0.429</td>\n",
       "      <td id=\"T_dd0ab_row6_col8\" class=\"data row6 col8\" >0.353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_dd0ab_level0_row7\" class=\"row_heading level0 row7\" >p110a</th>\n",
       "      <td id=\"T_dd0ab_row7_col0\" class=\"data row7 col0\" >0.500</td>\n",
       "      <td id=\"T_dd0ab_row7_col1\" class=\"data row7 col1\" >0.348</td>\n",
       "      <td id=\"T_dd0ab_row7_col2\" class=\"data row7 col2\" >0.381</td>\n",
       "      <td id=\"T_dd0ab_row7_col3\" class=\"data row7 col3\" >0.571</td>\n",
       "      <td id=\"T_dd0ab_row7_col4\" class=\"data row7 col4\" >0.533</td>\n",
       "      <td id=\"T_dd0ab_row7_col5\" class=\"data row7 col5\" >0.375</td>\n",
       "      <td id=\"T_dd0ab_row7_col6\" class=\"data row7 col6\" >0.429</td>\n",
       "      <td id=\"T_dd0ab_row7_col7\" class=\"data row7 col7\" >0.000</td>\n",
       "      <td id=\"T_dd0ab_row7_col8\" class=\"data row7 col8\" >0.474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_dd0ab_level0_row8\" class=\"row_heading level0 row8\" >p38a</th>\n",
       "      <td id=\"T_dd0ab_row8_col0\" class=\"data row8 col0\" >0.333</td>\n",
       "      <td id=\"T_dd0ab_row8_col1\" class=\"data row8 col1\" >0.278</td>\n",
       "      <td id=\"T_dd0ab_row8_col2\" class=\"data row8 col2\" >0.174</td>\n",
       "      <td id=\"T_dd0ab_row8_col3\" class=\"data row8 col3\" >0.258</td>\n",
       "      <td id=\"T_dd0ab_row8_col4\" class=\"data row8 col4\" >0.043</td>\n",
       "      <td id=\"T_dd0ab_row8_col5\" class=\"data row8 col5\" >0.190</td>\n",
       "      <td id=\"T_dd0ab_row8_col6\" class=\"data row8 col6\" >0.353</td>\n",
       "      <td id=\"T_dd0ab_row8_col7\" class=\"data row8 col7\" >0.474</td>\n",
       "      <td id=\"T_dd0ab_row8_col8\" class=\"data row8 col8\" >0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x172761580>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show matrix with background gradient\n",
    "cm = sns.light_palette(\"green\", as_cmap=True)\n",
    "kinase_distance_matrix_df.style.background_gradient(cmap=cm).format(\"{:.3f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: Since this is a distance matrix, lighter colors indicate similarity, darker colors dissimilarity."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Save kinase distance matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "kinase_distance_matrix_df.to_csv(DATA / \"kinase_distance_matrix.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Discussion\n",
    "\n",
    "In this talktorial, we have used the KLIFS interaction fingerprints, which describe binding modes of co-crystallized ligands in kinase structures, to assess kinase similarity.\n",
    "\n",
    "We have to keep two elements in mind:\n",
    "\n",
    "- Interaction fingerprints may miss important key interactions if the co-crystallized ligand(s) simply do not form the interaction but other ligands might.\n",
    "- We only compare here the two closest binding modes per kinase pair, although we have &mdash; for kinases such as EGFR and CDK2 &mdash; much more data available on binding modes. As an alternative, one could instead aggregate the information from multiple binding modes per kinase in a fingerprint that reports all interactions seen in any structure. However, this approach has its own drawbacks: Some kinases have much higher coverage than others, leading to an imbalance in information content.\n",
    "\n",
    "The kinase distance matrix above will be reloaded in __Talktorial T028__, where we compare kinase similarities from different perspectives, including the interaction perspective we have talked about in this talktorial."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quiz\n",
    "\n",
    "1. What are the advantages and disadvantages of describing kinase similarity with interaction fingerprints?\n",
    "2. Why is the structural coverage between kinases so different?\n",
    "3. At one point, we are mapping structure pairs to kinase pairs. We use here the minimum structure distance. What other mapping strategies can you think of? What advantages and disadvantages could they have?"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "toc-autonumbering": true,
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
